### ISSUE TITLE
Solaris 11 Git CLI error: Low level API call to digest SHA1 forbidden in FIPS mode!

### ISSUE THREAD

**Issue by markkvba** [oracle/solaris-openssl-fips#1]
Hello Chris/ Oracle. 

We are trying to setup a development box with jenkins/git on Solaris 11.  Jenkins Git plugin is throwing an error shown below.  Is there any anyway to make this work short of disabling FIPS mode?  I have seen some closed git issues stating SHA-1 is absolutely needed for GIT to work, so not going to be changed.   This is not a production system, but not sure IO would allow disabling FIPS, however, a reference to doing so would be appreciated. Any recommendations to make this work or punt? 

sha_locl.h(128): OpenSSL internal error, assertion failed: Low level API call to digest SHA1 forbidden in FIPS mode!
error: index-pack died of signal 6
fatal: index-pack failed

Regards,
Mark

**Comment by markkvba** [comment]
Created local copy of /etc/openssl/openssl.cnf with adding below sections and set  OPENSSL_CONF global environment var set in Jenkins, but this not appear to have any effect.

openssl_conf = openssl_options
[ openssl_options ]
alg_section = algs
[ algs ]
fips_mode = no

**Comment by vladak** [comment]
This issue is almost one year old however for reference purposes it still makes sense to provide an answer, I think.

Looking at the Jenkins Git plugin source code, it seems to be using both JGit implementation as well as the `git` command line tool: https://github.com/jenkinsci/git-plugin/blob/775c3d6754c5dee275d954392c6beb6e27fcba07/src/main/java/hudson/plugins/git/GitSCM.java#L1030

You're right that SHA1 for Git is absolutely vital, after all the commit IDs are SHA1 digests. That said, there is a project to move Git to algorithms that are not as susceptible to collisions: https://git-scm.com/docs/hash-function-transition/2.23.0

Back to the symptoms: the `Low level API call to digest SHA1 forbidden in FIPS mode!` error message comes from OpenSSL itself, specifically it is emitted during run time when someone calls one of the low level digest APIs directly when running OpenSSL in FIPS mode, i.e. calling `SHA1_Init()` in this case.

The other part of the error message seems to be coming from the [`git index-pack`](https://git-scm.com/docs/git-index-pack)  command. This makes sense given that the Jenkins Git plugin uses `git` command for some tasks. Looking at the `git` binary shipped with Solaris, it does link to OpenSSL:
```
$ ldd /usr/bin/git
        libpcre2-8.so.0 =>       /usr/lib/64/libpcre2-8.so.0
        libz.so.1 =>     /lib/64/libz.so.1
        libcrypto.so.1.0.0 =>    /lib/64/libcrypto.so.1.0.0
        libc.so.1 =>     /lib/64/libc.so.1
```
and it does call the low level SHA1 functions:
```
$ nm /usr/bin/git | grep SHA1_
[5397]  |          4299634592|                   0|FUNC |GLOB |0    |UNDEF  |SHA1_Final
[6672]  |          4299634656|                   0|FUNC |GLOB |0    |UNDEF  |SHA1_Init
[5352]  |          4299634624|                   0|FUNC |GLOB |0    |UNDEF  |SHA1_Update
```
so it is bound to fail when running OpenSSL in FIPS mode.

So, to piece it together, Jenkins Git plugin executes `git` command which calls `SHA1_Init()` which is forbidden in FIPS mode and hence the OpenSSL library shoots the process down with `SIGABRT` signal.

In fact, when I tried to use `git` on system running OpenSSL in FIPS mode, even the very basic workflow fails:
```
root@S11-4-SRU:~# openssl version
OpenSSL 1.0.2y-fips  16 Feb 2021
root@S11-4-SRU:~# cd /tmp/
root@S11-4-SRU:/tmp# mkdir gg
root@S11-4-SRU:/tmp# cd gg
root@S11-4-SRU:/tmp/gg# git init
...
root@S11-4-SRU:/tmp/gg# date > foo.txt
root@S11-4-SRU:/tmp/gg# git add foo.txt 
sha_locl.h(128): OpenSSL internal error, assertion failed: Low level API call to digest SHA1 forbidden in FIPS mode!
Abort (core dumped)
```

There are some possibilities:
  1. do not run OpenSSL in FIPS mode
    - to switch Solaris 11 system to (not) use OpenSSL in FIPS 140-2 mode is described in the official documentation on 
[Example of Running in FIPS 140-2 Mode on an Oracle Solaris 11.4 System](https://docs.oracle.com/cd/E37838_01/html/E61028/fips-example-s114.html#scrolltoc)
  1. fix Git to use the high level (i.e. `EVP_`) OpenSSL API to compute the digests
    - as a workaround, Git can be compiled in a way to avoid using SHA1 from OpenSSL (https://github.com/git/git/blob/0785eb769886ae81e346df10e88bc49ffc0ac64e/Makefile#L173-L174) however this is probably not something we would like to do in Solaris as it links to OpenSSL to use SHA1 on purpose (https://github.com/oracle/solaris-userland/blob/3ac9d6e3b7f12a9dbdf265be0e6bd3bdd1da5130/components/git/Makefile#L72-L78). Further, the SHA1 implementation built into Git is most probably not FIPS certified. That said, perhaps a bug can be filed against Git in Solaris to patch the Git source to use the `EVP_` API before it gets built. Or just compile your own Git.
  1. fix Jenkins Git plugin to avoid using the `git` command and rely just on JGit
    - JGit uses `java.security.MessageDigest` for digest computation (https://git.eclipse.org/r/plugins/gitiles/jgit/jgit/+/refs/heads/master/org.eclipse.jgit/src/org/eclipse/jgit/lib/Constants.java#492) so even if this ended up in Solaris Crypto Framework (via the Java PKCS#11 provider) running in FIPS mode, that would probably be fine.

**Comment by vladak** [comment]
There is another angle to this of course : why does OpenSSL disable access to SHA-1 low level APIs in FIPS mode given this is FIPS approved algorithm ? I don't want to spend too much on this here however as Darren Moffat confirmed internally this is because of OpenSSL 1.x layering design. This problem should not be present in OpenSSL 3.x (which is going to be FIPS certified eventually).

**Comment by vladak** [comment]
Also, for completeness: I tried converting the Git shipped with Solaris to OpenSSL 1.1.1 (which would solve the problem because 1.1.1 does not have FIPS mode) and this cannot be done in reasonable way currently as the straightforward solution would lead to `git-imap-send` to link to both OpenSSL 1.0 and 1.1.1 which is doomed to fail (the binary links with `libcurl` that depends on OpenSSL 1.0) and I don't really want to patch the Git build excessively. A workaround might be to run `git` with `LD_LIBRARY_PATH=/lib/openssl/default/64`, i.e. point the dynamic linker to the non-FIPS version of OpenSSL however as the internet warns, such things are ["evil"](http://xahlee.info/UnixResource_dir/_/ldpath.html).
